### PR TITLE
Improve 003 benchmark

### PR DESCRIPTION
**Pull request by sscargal**

### CODE REVIEW

**Comment by sscargal** [comment]
I tested the benchmark on a 2-Socket Supermicro server with 2 x Intel Gold 6438Y+ (32-core) CPUs and an NVidia A10.

Since my A10 GPU is on a PCIe slot for CPU Socket 1, I used `ndctl` to ensure the benchmark runs on Socket1 using memory from Socket1 to avoid cross-socket communication that will impact the results. I ran the benchmark several times to look for consistency. The first result is slower than the rest, likely due to page faults.

```bash
$ for f in {1..10}; do numactl --cpunodebind=1 --membind=1 ./main; done
Bandwidth: 14.857154 GB/s
Bandwidth: 18.323864 GB/s
Bandwidth: 19.694246 GB/s
Bandwidth: 19.409304 GB/s
Bandwidth: 18.445518 GB/s
Bandwidth: 19.691296 GB/s
Bandwidth: 19.410233 GB/s
Bandwidth: 19.578987 GB/s
Bandwidth: 19.683897 GB/s
Bandwidth: 19.635284 GB/s
```

This ~20GB/sec is lower than I expected. The theoretical bi-directional throughput of a x16 PCIe Gen4 slot is 32GB/s.

I then read the [NVidia CUDA Best Practices](https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html) documentation. Reviewing the code showed several areas for improvement:

- **Single-threaded Execution:** The code is single-threaded and uses a single CUDA stream to perform data transfers. While this is a common approach, it may not fully utilize the available bandwidth of PCIe Gen4.
- **Asynchronous Transfers:** The code uses cudaMemcpyAsync for asynchronous data transfers, which is good for overlapping computation and communication. However, since the code waits for the stream to synchronize after recording the stop event, it effectively serializes the transfer operations, which might not fully leverage the potential parallelism.
- **Memory Allocation:** The host memory is allocated using standard dynamic memory allocation, which may not be page-locked (pinned). Using pinned memory can significantly increase transfer rates because it allows for faster data transfers between the host and device.
- **No Warmup:** Data is immediately allocated and used. Benchmarks typically perform some operations before measurements start to avoid page faults on first use, which can impact the results.

I modified the code to include these observations and got 25GB/s, which is close to my expectation. We won't get 32GB/sec, but this is "close enough".

```bash
$ for f in {1..10}; do numactl --cpunodebind=1 --membind=1 ./main -i 2000 -s 8 -w 5; done
Bandwidth: 25.149124 GB/s
Bandwidth: 25.151405 GB/s
Bandwidth: 25.150078 GB/s
Bandwidth: 25.150782 GB/s
Bandwidth: 25.147242 GB/s
Bandwidth: 25.148067 GB/s
Bandwidth: 25.147434 GB/s
Bandwidth: 25.147575 GB/s
Bandwidth: 25.147165 GB/s
Bandwidth: 25.146469 GB/s
```

Note: This is the same result I get from the [NVidia nvbandwidth benchmark](https://github.com/NVIDIA/nvbandwidth) available on GitHub, which gives me confidence my changes were successful.

```bash
$ ./nvbandwidth -t host_to_device_memcpy_ce
nvbandwidth Version: v0.5
Built from Git version: v0.5-1-g4da7d7e

NOTE: This tool reports current measured bandwidth on your system.
Additional system-specific tuning may be required to achieve maximal peak bandwidth.

CUDA Runtime Version: 12040
CUDA Driver Version: 12040
Driver Version: 550.54.15

Device 0: NVIDIA A10

Running host_to_device_memcpy_ce.
memcpy CE CPU(row) -> GPU(column) bandwidth (GB/s)
           0
 0     25.17

SUM host_to_device_memcpy_ce 25.17
```

```bash
$ ./nvbandwidth -t host_to_device_memcpy_sm
nvbandwidth Version: v0.5
Built from Git version: v0.5-1-g4da7d7e

NOTE: This tool reports current measured bandwidth on your system.
Additional system-specific tuning may be required to achieve maximal peak bandwidth.

CUDA Runtime Version: 12040
CUDA Driver Version: 12040
Driver Version: 550.54.15

Device 0: NVIDIA A10

Running host_to_device_memcpy_sm.
memcpy SM CPU(row) -> GPU(column) bandwidth (GB/s)
           0
 0     25.25

SUM host_to_device_memcpy_sm 25.25
```

HTH

File: 003_bandwidth_check/Makefile [hunk 1]
```diff
@@ -1,9 +1,9 @@
 all: hw
 
 hw:
-	 nvcc -x cu -O3 -std=c++11 -arch=sm_86 -I/usr/local/cuda-12.3/bin/../include ./bandwidth_check.cu -o main -L/usr/local/cuda-12.3/bin/../lib64
+	 nvcc -x cu -O3 -std=c++11 -arch=sm_86 -I/usr/local/cuda-12.3/bin/../include ./bandwidth_check.cu -o bandwidth_check -L/usr/local/cuda-12.3/bin/../lib64
```
Comment (drkennetz) [inline 003_bandwidth_check/Makefile:4]: Please change this back, I am using a convention to make every executable "main" so it is easy from a gitignore perspective.

File: 003_bandwidth_check/bandwidth_check.cu [hunk 2]
```diff
@@ -1,49 +1,128 @@
...
+
     // Calculate the elapsed time.
     float ms = 0;
     ::cudaEventElapsedTime(&ms, start, stop);
-    // Convert to GB/s.
-    const float bandwidth = ((float(dataSize) * sizeof(float) * float(NUM_COPY_ITERATIONS)) / (ms * 1e6));
-    printf("Bandwidth: %f GB/s\n", bandwidth);
+
+    // Check if ms is zero to avoid division by zero.
+    if (ms > 0) {
```
Comment (drkennetz) [inline 003_bandwidth_check/bandwidth_check.cu:72]: Nice check

File: 003_bandwidth_check/bandwidth_check.cu [hunk 3]
```diff
@@ -1,49 +1,128 @@
...
+// Default values
+constexpr int DEFAULT_NUM_COPY_ITERATIONS = 1000;
+constexpr int DEFAULT_NUM_STREAMS = 4;
+constexpr int DEFAULT_WARMUP_ITERATIONS = 10;
+constexpr size_t DEFAULT_DATA_SIZE = 1 << 24; // 16M floats
+
+void checkBandwidth(size_t dataSize, int numCopyIterations, int numStreams, int warmupIterations) {
+    // Allocate pinned host memory.
+    float* hData;
+    ::cudaHostAlloc(&hData, dataSize * sizeof(float), cudaHostAllocDefault);
```
Comment (drkennetz) [inline 003_bandwidth_check/bandwidth_check.cu:16]: Good call on host pinned memory

File: 003_bandwidth_check/bandwidth_check.cu [hunk 4]
```diff
@@ -1,49 +1,128 @@
...
+
     // Create CUDA events for timing purposes.
     cudaEvent_t start, stop;
     ::cudaEventCreate(&start);
     ::cudaEventCreate(&stop);
+
+    // Create multiple CUDA streams.
+    std::vector<cudaStream_t> streams(numStreams);
+    for (int i = 0; i < numStreams; ++i) {
+        ::cudaStreamCreate(&streams[i]);
```
Comment (drkennetz) [inline 003_bandwidth_check/bandwidth_check.cu:35]: I don't see why this would make a difference. Did you try it on a single stream?

**Review by drkennetz — APPROVED** [review]

**Comment by drkennetz** [comment]
@sscargal thanks for contributing, really nice changes.

**Pull request merged**